File: python/proxy.py
import socket
import threading
import os
import time
import select
import requests
import pickle
import logging
from datetime import datetime
from concurrent.futures import ThreadPoolExecutor
from urllib.parse import urlparse, urljoin

# Constants
CACHE_DIR = "cache"
LOGS_DIR = "logs"
BUFFER_SIZE = 8192
MAX_WORKERS = 50
DEFAULT_TIMEOUT = 30

class ProxyLogger:
    def __init__(self):
        # Create logs directory if it doesn't exist
        os.makedirs(LOGS_DIR, exist_ok=True)
        
        # Set up main logger
        self.logger = logging.getLogger('ProxyServer')
        self.logger.setLevel(logging.INFO)
        
        # Create handlers for different log types
        self._setup_access_logger()
        self._setup_error_logger()
        self._setup_security_logger()
        
        # Traffic statistics
        self.stats = {
            'total_requests': 0,
            'blocked_requests': 0,
            'bytes_transferred': 0
        }
    
    def _setup_access_logger(self):
        # Access log format: timestamp | client_ip | method | url | status_code | bytes
        access_handler = logging.FileHandler(os.path.join(LOGS_DIR, 'access.log'))
        access_handler.setLevel(logging.INFO)
        access_formatter = logging.Formatter('%(asctime)s | %(message)s')
        access_handler.setFormatter(access_formatter)
        
        self.access_logger = logging.getLogger('ProxyServer.access')
        self.access_logger.setLevel(logging.INFO)
        self.access_logger.addHandler(access_handler)
    
    def _setup_error_logger(self):
        error_handler = logging.FileHandler(os.path.join(LOGS_DIR, 'error.log'))
        error_handler.setLevel(logging.ERROR)
        error_formatter = logging.Formatter('%(asctime)s | %(levelname)s | %(message)s')
        error_handler.setFormatter(error_formatter)
        
        self.error_logger = logging.getLogger('ProxyServer.error')
        self.error_logger.setLevel(logging.ERROR)
        self.error_logger.addHandler(error_handler)
    
    def _setup_security_logger(self):
        security_handler = logging.FileHandler(os.path.join(LOGS_DIR, 'security.log'))
        security_handler.setLevel(logging.WARNING)
        security_formatter = logging.Formatter('%(asctime)s | SECURITY | %(message)s')
        security_handler.setFormatter(security_formatter)
        
        self.security_logger = logging.getLogger('ProxyServer.security')
        self.security_logger.setLevel(logging.WARNING)
        self.security_logger.addHandler(security_handler)
    
    def log_access(self, client_ip, method, url, status_code, bytes_transferred):
        self.stats['total_requests'] += 1
        self.stats['bytes_transferred'] += bytes_transferred
        self.access_logger.info(f"{client_ip} | {method} | {url} | {status_code} | {bytes_transferred}")
    
    def log_error(self, client_ip, message, exception=None):
        error_msg = f"{client_ip} | {message}"
        if exception:
            error_msg += f" | Exception: {str(exception)}"
        self.error_logger.error(error_msg)
    
    def log_security(self, client_ip, message):
        self.stats['blocked_requests'] += 1
        self.security_logger.warning(f"{client_ip} | {message}")
    
    def get_statistics(self):
        return {
            'total_requests': self.stats['total_requests'],
            'blocked_requests': self.stats['blocked_requests'],
            'bytes_transferred': self.stats['bytes_transferred'],
            'bytes_transferred_formatted': self._format_bytes(self.stats['bytes_transferred'])
        }
    
    def _format_bytes(self, bytes):
        for unit in ['B', 'KB', 'MB', 'GB']:
            if bytes < 1024:
                return f"{bytes:.2f} {unit}"
            bytes /= 1024
        return f"{bytes:.2f} TB"

class Proxy:
    def __init__(self, port):
        self.port = port
        self.server_socket = None
        self.run = True
        self.cache_file = {}
        self.block_sites = {}
        self.servicing_threads = []
        self.logger = ProxyLogger()
        
        # Ensure directories exist
        os.makedirs(CACHE_DIR, exist_ok=True)
        
        # Load cached and blocked sites
        self.load_cache()
        self.load_blocked_sites()
        
        # Initialize thread pool
        self.thread_pool = ThreadPoolExecutor(max_workers=MAX_WORKERS)
        
        # Start management interface
        threading.Thread(target=self.dynamic_manager, daemon=True).start()

    def load_cache(self):
        try:
            if os.path.exists("cached_sites.pkl"):
                with open("cached_sites.pkl", "rb") as f:
                    self.cache_file = pickle.load(f)
        except Exception as e:
            self.logger.log_error("system", "Cache loading error", e)
            self.cache_file = {}

    def load_blocked_sites(self):
        try:
            if os.path.exists("blocked_sites.pkl"):
                with open("blocked_sites.pkl", "rb") as f:
                    self.block_sites = pickle.load(f)
        except Exception as e:
            self.logger.log_error("system", "Blocked sites loading error", e)
            self.block_sites = {}

    def save_cache(self):
        with open("cached_sites.pkl", "wb") as f:
            pickle.dump(self.cache_file, f)

    def save_blocked_sites(self):
        with open("blocked_sites.pkl", "wb") as f:
            pickle.dump(self.block_sites, f)

    def listen(self):
        try:
            self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.server_socket.bind(("", self.port))
            self.server_socket.listen(100)
            print(f"Proxy server listening on port {self.port}")

            while self.run:
                try:
                    client_socket, addr = self.server_socket.accept()
                    client_socket.settimeout(DEFAULT_TIMEOUT)
                    print(f"Connection from {addr}")
                    self.thread_pool.submit(self.handle_client, client_socket, addr[0])
                except Exception as e:
                    self.logger.log_error("system", "Error accepting connection", e)

        except Exception as e:
            self.logger.log_error("system", "Server error", e)
        finally:
            self.close_server()

    def handle_client(self, client_socket, client_ip):
        try:
            handler = RequestHandler(client_socket, self, client_ip)
            handler.handle_request()
        except Exception as e:
            self.logger.log_error(client_ip, "Error handling client", e)
        finally:
            try:
                client_socket.close()
            except:
                pass

    def close_server(self):
        print("Shutting down proxy server...")
        self.run = False
        self.save_cache()
        self.save_blocked_sites()
        self.thread_pool.shutdown(wait=True)
        
        if self.server_socket:
            try:
                self.server_socket.close()
            except:
                pass

    def dynamic_manager(self):
        while self.run:
            try:
                print("\nProxy Menu:")
                print("1. Block a site")
                print("2. View blocked sites")
                print("3. View cached sites")
                print("4. View statistics")
                print("5. Clear cache")
                print("6. View recent logs")
                print("7. Exit")
                
                choice = input("Enter choice (1-7): ").strip()
                
                if choice == "1":
                    site = input("Enter site to block: ").strip()
                    self.block_sites[site] = True
                    print(f"Blocked {site}")
                    self.save_blocked_sites()
                elif choice == "2":
                    print("Blocked sites:", list(self.block_sites.keys()))
                elif choice == "3":
                    print("Cached sites:", list(self.cache_file.keys()))
                elif choice == "4":
                    stats = self.logger.get_statistics()
                    print("\nProxy Statistics:")
                    print(f"Total Requests: {stats['total_requests']}")
                    print(f"Blocked Requests: {stats['blocked_requests']}")
                    print(f"Total Data Transferred: {stats['bytes_transferred_formatted']}")
                elif choice == "5":
                    self.cache_file.clear()
                    self.save_cache()
                    print("Cache cleared")
                elif choice == "6":
                    self.show_recent_logs()
                elif choice == "7":
                    self.close_server()
                    break
            except Exception as e:
                self.logger.log_error("system", "Menu error", e)

    def show_recent_logs(self, lines=10):
        print("\nRecent Access Logs:")
        self._show_last_lines(os.path.join(LOGS_DIR, 'access.log'), lines)
        
        print("\nRecent Error Logs:")
        self._show_last_lines(os.path.join(LOGS_DIR, 'error.log'), lines)
        
        print("\nRecent Security Logs:")
        self._show_last_lines(os.path.join(LOGS_DIR, 'security.log'), lines)

    def _show_last_lines(self, filename, lines):
        try:
            if os.path.exists(filename):
                with open(filename, 'r') as f:
                    content = f.readlines()
                    for line in content[-lines:]:
                        print(line.strip())
            else:
                print(f"No logs found in {filename}")
        except Exception as e:
            print(f"Error reading logs: {e}")

class RequestHandler:
    def __init__(self, client_socket, proxy, client_ip):
        self.client_socket = client_socket
        self.proxy = proxy
        self.client_ip = client_ip
        self.request_data = b""
        self.headers = {}
        self.bytes_transferred = 0

    def handle_request(self):
        try:
            # Receive initial request
            while b"\r\n\r\n" not in self.request_data:
                data = self.client_socket.recv(BUFFER_SIZE)
                if not data:
                    return
                self.request_data += data

            # Parse the request
            request_lines = self.request_data.split(b"\r\n")
            request_line = request_lines[0].decode()
            
            try:
                method, url, version = request_line.split(" ")
            except ValueError:
                self.proxy.logger.log_error(self.client_ip, f"Invalid request line: {request_line}")
                return

            # Parse headers
            for line in request_lines[1:]:
                try:
                    if line and b":" in line:
                        key, value = line.decode().split(":", 1)
                        self.headers[key.strip().lower()] = value.strip()
                except Exception as e:
                    self.proxy.logger.log_error(self.client_ip, "Header parsing error", e)

            # Handle different request types
            if method == "CONNECT":
                self.handle_https_request(url)
            else:
                self.handle_http_request(method, url)

        except Exception as e:
            self.proxy.logger.log_error(self.client_ip, "Request handling error", e)

    def handle_http_request(self, method, url):
        try:
            # Ensure URL is absolute
            if not url.startswith('http'):
                host = self.headers.get('host', '')
                url = f'http://{host}{url}'

            # Check if site is blocked
            parsed_url = urlparse(url)
            if parsed_url.netloc in self.proxy.block_sites:
                self.proxy.logger.log_security(self.client_ip, f"Blocked access to {url}")
                self.send_blocked_response()
                return

            # Create request
            request_headers = {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
                'Accept': '*/*',
                'Accept-Encoding': 'identity',
                'Connection': 'close'
            }

            # Add original headers
            for key, value in self.headers.items():
                if key not in ['proxy-connection', 'connection']:
                    request_headers[key] = value

            # Make request
            response = requests.request(
                method=method,
                url=url,
                headers=request_headers,
                stream=True,
                verify=False,
                allow_redirects=True
            )

            # Send response headers
            header_string = f"HTTP/{response.raw.version / 10.0} {response.status_code} {response.reason}\r\n"
            for key, value in response.headers.items():
                header_string += f"{key}: {value}\r\n"
            header_string += "\r\n"
            
            self.client_socket.sendall(header_string.encode())

            # Stream response body
            for chunk in response.iter_content(chunk_size=BUFFER_SIZE):
                if chunk:
                    self.client_socket.sendall(chunk)
                    self.bytes_transferred += len(chunk)

            # Log the successful request
            self.proxy.logger.log_access(
                self.client_ip,
                method,
                url,
                response.status_code,
                self.bytes_transferred
            )

        except Exception as e:
            self.proxy.logger.log_error(self.client_ip, f"HTTP request error for {url}", e)
            self.send_error_response(502)

    def handle_https_request(self, url):
        try:
            host, port = url.split(':')
            port = int(port)

            # Log HTTPS connection attempt
            self.proxy.logger.log_access(self.client_ip, "CONNECT", url, 200, 0)

            # Create connection to remote server
            server_socket = socket.create_connection((host, port), timeout=DEFAULT_TIMEOUT)
            
            # Send connection established response
            self.client_socket.sendall(b"HTTP/1.1 200 Connection Established\r\n\r\n")

            # Set up bidirectional tunneling
            self.tunnel_traffic(self.client_socket, server_socket)

        except Exception as e:
            self.proxy.logger.log_error(self.client_ip, f"HTTPS request error for {url}", e)
            self.send_error_response(502)

    def tunnel_traffic(self, client_socket, server_socket):
        try:
            sockets = [client_socket, server_socket]
            timeout = 1
            while True:
                readable, _, exceptional = select.select(sockets, [], sockets, timeout)

                if exceptional:
                    break

                for sock in readable:
                    other = server_socket if sock is client_socket else client_socket
                    try:
                        data = sock.recv(BUFFER_SIZE)
                        if not data:
                            return
                        other.sendall(data)
                    except Exception as e:
                        print(f"Tunneling error: {e}")
                        return

        except Exception as e:
            print(f"Tunnel error: {e}")
        finally:
            try:
                client_socket.close()
                server_socket.close()
            except:
                pass

    def send_blocked_response(self):
        response = (
            "HTTP/1.1 403 Forbidden\r\n"
            "Content-Type: text/html\r\n"
            "Connection: close\r\n"
            "\r\n"
            "<html><body><h1>403 Forbidden</h1><p>This site is blocked by the proxy.</p></body></html>"
        )
        self.client_socket.sendall(response.encode())

    def send_error_response(self, status_code):
        response = (
            f"HTTP/1.1 {status_code} Error\r\n"
            "Content-Type: text/html\r\n"
            "Connection: close\r\n"
            "\r\n"
            f"<html><body><h1>{status_code} Error</h1></body></html>"
        )
        try:
            self.client_socket.sendall(response.encode())
        except:
            pass

if __name__ == "__main__":
    proxy = Proxy(8080)
    try:
        proxy.listen()
    except KeyboardInterrupt:
        proxy.close_server()

File: cpp/Proxy.cpp
#include <iostream>
#include <string>
#include <thread>
#include <vector>
#include <map>
#include <mutex>
#include <fstream>
#include <sstream>
#include <chrono>
#include <filesystem>
#include <ctime>
#include <memory>
#include <algorithm>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <poll.h>
#include <signal.h>

// Constants
constexpr int BUFFER_SIZE = 8192;
constexpr int MAX_WORKERS = 50;
constexpr int DEFAULT_TIMEOUT = 30;
const std::string CACHE_DIR = "cache";
const std::string LOGS_DIR = "logs";

class ProxyLogger {
private:
    std::mutex log_mutex;
    std::ofstream access_log;
    std::ofstream error_log;
    std::ofstream security_log;
    
    struct Statistics {
        size_t total_requests{0};
        size_t blocked_requests{0};
        size_t bytes_transferred{0};
    } stats;

    std::string format_bytes(size_t bytes) {
        const char* units[] = {"B", "KB", "MB", "GB", "TB"};
        int unit_index = 0;
        double size = static_cast<double>(bytes);
        
        while (size >= 1024 && unit_index < 4) {
            size /= 1024;
            unit_index++;
        }
        
        std::stringstream ss;
        ss << std::fixed << std::setprecision(2) << size << " " << units[unit_index];
        return ss.str();
    }

    void setup_log_files() {
        std::filesystem::create_directories(LOGS_DIR);
        access_log.open(LOGS_DIR + "/access.log", std::ios::app);
        error_log.open(LOGS_DIR + "/error.log", std::ios::app);
        security_log.open(LOGS_DIR + "/security.log", std::ios::app);
    }

public:
    ProxyLogger() {
        setup_log_files();
    }

    ~ProxyLogger() {
        access_log.close();
        error_log.close();
        security_log.close();
    }

    void log_access(const std::string& client_ip, const std::string& method,
                   const std::string& url, int status_code, size_t bytes_transferred) {
        std::lock_guard<std::mutex> lock(log_mutex);
        stats.total_requests++;
        stats.bytes_transferred += bytes_transferred;
        
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        
        access_log << std::ctime(&time) << client_ip << " | " 
                  << method << " | " << url << " | " 
                  << status_code << " | " << bytes_transferred << std::endl;
    }

    void log_error(const std::string& client_ip, const std::string& message,
                  const std::exception* e = nullptr) {
        std::lock_guard<std::mutex> lock(log_mutex);
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        
        error_log << std::ctime(&time) << client_ip << " | " << message;
        if (e) {
            error_log << " | Exception: " << e->what();
        }
        error_log << std::endl;
    }

    void log_security(const std::string& client_ip, const std::string& message) {
        std::lock_guard<std::mutex> lock(log_mutex);
        stats.blocked_requests++;
        
        auto now = std::chrono::system_clock::now();
        auto time = std::chrono::system_clock::to_time_t(now);
        
        security_log << std::ctime(&time) << "SECURITY | " 
                    << client_ip << " | " << message << std::endl;
    }

    Statistics get_statistics() const {
        return stats;
    }
};

class RequestHandler;

class Proxy {
private:
    int port;
    int server_socket;
    bool running;
    std::map<std::string, std::string> cache_file;
    std::map<std::string, bool> block_sites;
    std::vector<std::thread> worker_threads;
    ProxyLogger logger;
    std::mutex cache_mutex;
    std::mutex sites_mutex;

    // void load_cache() {
    //     try {
    //         std::ifstream file("cached_sites.dat", std::ios::binary);
    //         if (file.is_open()) {
    //             // Implement cache loading logic
    //             file.close();
    //         }
    //     } catch (const std::exception& e) {
    //         logger.log_error("system", "Cache loading error", &e);
    //     }
    // }

    void load_blocked_sites() {
        try {
            std::ifstream file("blocked_sites.dat", std::ios::binary);
            if (file.is_open()) {
                // Implement blocked sites loading logic
                file.close();
            }
        } catch (const std::exception& e) {
            logger.log_error("system", "Blocked sites loading error", &e);
        }
    }

    // void save_cache() {
    //     std::lock_guard<std::mutex> lock(cache_mutex);
    //     std::ofstream file("cached_sites.dat", std::ios::binary);
    //     // Implement cache saving logic
    // }

    void save_blocked_sites() {
        std::lock_guard<std::mutex> lock(sites_mutex);
        std::ofstream file("blocked_sites.dat", std::ios::binary);
        // Implement blocked sites saving logic
    }

public:
    Proxy(int port_) : port(port_), running(true) {
        std::filesystem::create_directories(CACHE_DIR);
        load_cache();
        load_blocked_sites();
        
        // Handle SIGINT gracefully
        signal(SIGINT, [](int) {
            // Cleanup code here
            exit(0);
        });
    }

    ~Proxy() {
        stop();
    }

    void start() {
        server_socket = socket(AF_INET, SOCK_STREAM, 0);
        if (server_socket < 0) {
            throw std::runtime_error("Failed to create socket");
        }

        int opt = 1;
        setsockopt(server_socket, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

        sockaddr_in server_addr{};
        server_addr.sin_family = AF_INET;
        server_addr.sin_addr.s_addr = INADDR_ANY;
        server_addr.sin_port = htons(port);

        if (bind(server_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
            throw std::runtime_error("Failed to bind socket");
        }

        if (listen(server_socket, 100) < 0) {
            throw std::runtime_error("Failed to listen on socket");
        }

        std::cout << "Proxy server listening on port " << port << std::endl;

        // Start management interface in a separate thread
        std::thread mgmt_thread(&Proxy::dynamic_manager, this);
        mgmt_thread.detach();

        // Main accept loop
        while (running) {
            sockaddr_in client_addr{};
            socklen_t client_len = sizeof(client_addr);
            
            int client_socket = accept(server_socket, (struct sockaddr*)&client_addr, &client_len);
            if (client_socket < 0) {
                logger.log_error("system", "Failed to accept connection");
                continue;
            }

            std::string client_ip = inet_ntoa(client_addr.sin_addr);
            std::cout << "Connection from " << client_ip << std::endl;

            // Create new thread for client handling
            worker_threads.emplace_back(&Proxy::handle_client, this, client_socket, client_ip);
        }
    }

    void stop() {
        running = false;
        save_cache();
        save_blocked_sites();
        
        // Join all worker threads
        for (auto& thread : worker_threads) {
            if (thread.joinable()) {
                thread.join();
            }
        }
        
        if (server_socket >= 0) {
            close(server_socket);
        }
    }

    // void handle_client(int client_socket, const std::string& client_ip) {
    //     try {
    //         // Create request handler and process request
    //         RequestHandler handler(client_socket, this, client_ip, logger);
    //         handler.handle_request();
    //     } catch (const std::exception& e) {
    //         logger.log_error(client_ip, "Error handling client", &e);
    //     }
        
    //     close(client_socket);
    // }

    void dynamic_manager() {
        while (running) {
            std::cout << "\nProxy Menu:\n"
                     << "1. Block a site\n"
                     << "2. View blocked sites\n"
                     << "3. View cached sites\n"
                     << "4. View statistics\n"
                     << "5. Clear cache\n"
                     << "6. View recent logs\n"
                     << "7. Exit\n"
                     << "Enter choice (1-7): ";

            std::string choice;
            std::getline(std::cin, choice);

            try {
                if (choice == "1") {
                    std::cout << "Enter site to block: ";
                    std::string site;
                    std::getline(std::cin, site);
                    std::lock_guard<std::mutex> lock(sites_mutex);
                    block_sites[site] = true;
                    std::cout << "Blocked " << site << std::endl;
                    save_blocked_sites();
                }
                // Implement other menu options...
            } catch (const std::exception& e) {
                logger.log_error("system", "Menu error", &e);
            }
        }
    }

    bool is_site_blocked(const std::string& site) const {
        std::lock_guard<std::mutex> lock(sites_mutex);
        return block_sites.find(site) != block_sites.end();
    }
};

class RequestHandler {
private:
    int client_socket;
    Proxy* proxy;
    std::string client_ip;
    ProxyLogger& logger;
    std::vector<char> request_data;
    std::map<std::string, std::string> headers;
    size_t bytes_transferred;

    void parse_headers(const std::string& header_data) {
        std::istringstream stream(header_data);
        std::string line;
        
        while (std::getline(stream, line) && !line.empty()) {
            auto colon_pos = line.find(':');
            if (colon_pos != std::string::npos) {
                std::string key = line.substr(0, colon_pos);
                std::string value = line.substr(colon_pos + 1);
                
                // Trim whitespace
                key.erase(0, key.find_first_not_of(" \t"));
                key.erase(key.find_last_not_of(" \t") + 1);
                value.erase(0, value.find_first_not_of(" \t"));
                value.erase(value.find_last_not_of(" \t") + 1);
                
                headers[key] = value;
            }
        }
    }

public:
    RequestHandler(int socket, Proxy* proxy_, const std::string& ip, ProxyLogger& logger_)
        : client_socket(socket), proxy(proxy_), client_ip(ip), logger(logger_),
          bytes_transferred(0) {}

    void handle_request() {
        try {
            char buffer[BUFFER_SIZE];
            std::string header_data;
            bool headers_complete = false;
            
            while (!headers_complete) {
                ssize_t bytes_read = recv(client_socket, buffer, BUFFER_SIZE, 0);
                if (bytes_read <= 0) {
                    return;
                }
                
                header_data.append(buffer, bytes_read);
                if (header_data.find("\r\n\r\n") != std::string::npos) {
                    headers_complete = true;
                }
            }

            // Parse request line
            std::istringstream request_stream(header_data);
            std::string request_line;
            std::getline(request_stream, request_line);
            
            std::istringstream request_parts(request_line);
            std::string method, url, version;
            request_parts >> method >> url >> version;

            // Parse headers
            parse_headers(header_data.substr(request_line.length() + 2));

            // Handle request based on method
            if (method == "CONNECT") {
                handle_https_request(url);
            } else {
                handle_http_request(method, url);
            }

        } catch (const std::exception& e) {
            logger.log_error(client_ip, "Request handling error", &e);
        }
    }

    void handle_http_request(const std::string& method, const std::string& url) {
        // Implementation of HTTP request handling
        // This would involve creating a connection to the target server,
        // forwarding the request, and handling the response
    }

    void handle_https_request(const std::string& url) {
        // Implementation of HTTPS tunneling
        // This would involve creating an SSL tunnel between the client
        // and the target server
    }

    void send_blocked_response() {
        const char* response =
            "HTTP/1.1 403 Forbidden\r\n"
            "Content-Type: text/html\r\n"
            "Connection: close\r\n"
            "\r\n"
            "<html><body><h1>403 Forbidden</h1>"
            "<p>This site is blocked by the proxy.</p></body></html>";
            
        send(client_socket, response, strlen(response), 0);
    }

    void send_error_response(int status_code) {
        std::stringstream response;
        response << "HTTP/1.1 " << status_code << " Error\r\n"
                << "Content-Type: text/html\r\n"
                << "Connection: close\r\n"
                << "\r\n"
                << "<html><body><h1>" << status_code << " Error</h1></body></html>";
                
        std::string response_str = response.str();
        send(client_socket, response_str.c_str(), response_str.length(), 0);
    }
};

// Cache loading implementation
void Proxy::load_cache() {
    try {
        std::ifstream file("cached_sites.dat", std::ios::binary);
        if (file.is_open()) {
            std::lock_guard<std::mutex> lock(cache_mutex);
            size_t cache_size;
            file.read(reinterpret_cast<char*>(&cache_size), sizeof(size_t));
            
            for (size_t i = 0; i < cache_size; ++i) {
                size_t key_size, value_size;
                
                // Read key size and key
                file.read(reinterpret_cast<char*>(&key_size), sizeof(size_t));
                std::string key(key_size, '\0');
                file.read(&key[0], key_size);
                
                // Read value size and value
                file.read(reinterpret_cast<char*>(&value_size), sizeof(size_t));
                std::string value(value_size, '\0');
                file.read(&value[0], value_size);
                
                cache_file[key] = value;
            }
            file.close();
        }
    } catch (const std::exception& e) {
        logger.log_error("system", "Cache loading error", &e);
    }
}

// Save cache implementation
void Proxy::save_cache() {
    std::lock_guard<std::mutex> lock(cache_mutex);
    std::ofstream file("cached_sites.dat", std::ios::binary);
    
    if (file.is_open()) {
        // Write cache size
        size_t cache_size = cache_file.size();
        file.write(reinterpret_cast<const char*>(&cache_size), sizeof(size_t));
        
        // Write each cache entry
        for (const auto& entry : cache_file) {
            // Write key
            size_t key_size = entry.first.size();
            file.write(reinterpret_cast<const char*>(&key_size), sizeof(size_t));
            file.write(entry.first.c_str(), key_size);
            
            // Write value
            size_t value_size = entry.second.size();
            file.write(reinterpret_cast<const char*>(&value_size), sizeof(size_t));
            file.write(entry.second.c_str(), value_size);
        }
        file.close();
    }
}

// Complete handle_client implementation
void Proxy::handle_client(int client_socket, const std::string& client_ip) {
    try {
        char buffer[BUFFER_SIZE];
        std::string request_data;
        ssize_t bytes_received;
        
        // Set socket timeout
        struct timeval timeout;
        timeout.tv_sec = DEFAULT_TIMEOUT;
        timeout.tv_usec = 0;
        setsockopt(client_socket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
        
        // Receive request
        while ((bytes_received = recv(client_socket, buffer, BUFFER_SIZE - 1, 0)) > 0) {
            buffer[bytes_received] = '\0';
            request_data += buffer;
            if (request_data.find("\r\n\r\n") != std::string::npos) {
                break;
            }
        }

        if (bytes_received <= 0) {
            logger.log_error(client_ip, "Connection closed by client");
            close(client_socket);
            return;
        }

        // Parse request
        std::istringstream request_stream(request_data);
        std::string request_line;
        std::getline(request_stream, request_line);
        
        std::istringstream request_parts(request_line);
        std::string method, url, version;
        request_parts >> method >> url >> version;

        // Parse headers
        std::map<std::string, std::string> headers;
        std::string header_line;
        while (std::getline(request_stream, header_line) && header_line != "\r") {
            size_t colon_pos = header_line.find(':');
            if (colon_pos != std::string::npos) {
                std::string key = header_line.substr(0, colon_pos);
                std::string value = header_line.substr(colon_pos + 2); // Skip ": "
                if (!value.empty() && value.back() == '\r') {
                    value.pop_back();
                }
                headers[key] = value;
            }
        }

        // Handle request based on method
        if (method == "CONNECT") {
            handle_https_request(client_socket, url, headers, client_ip);
        } else {
            handle_http_request(client_socket, method, url, headers, request_data, client_ip);
        }

    } catch (const std::exception& e) {
        logger.log_error(client_ip, "Error handling client", &e);
    }
    
    close(client_socket);
}

// Utility function for tunneling data
void tunnel_data(int source_socket, int dest_socket, const std::string& client_ip,
                const std::string& direction, ProxyLogger& logger) {
    char buffer[BUFFER_SIZE];
    ssize_t bytes_read;

    while ((bytes_read = recv(source_socket, buffer, BUFFER_SIZE, 0)) > 0) {
        ssize_t bytes_sent = send(dest_socket, buffer, bytes_read, 0);
        if (bytes_sent <= 0) {
            logger.log_error(client_ip, "Failed to forward data: " + direction);
            break;
        }
    }

    if (bytes_read < 0) {
        logger.log_error(client_ip, "Error reading data: " + direction);
    }
}

// Helper function for HTTP response handling
void forward_response(int client_socket, int server_socket, const std::string& client_ip,
                     ProxyLogger& logger) {
    char buffer[BUFFER_SIZE];
    ssize_t bytes_received;
    size_t total_bytes = 0;

    while ((bytes_received = recv(server_socket, buffer, BUFFER_SIZE, 0)) > 0) {
        ssize_t bytes_sent = send(client_socket, buffer, bytes_received, 0);
        if (bytes_sent <= 0) {
            logger.log_error(client_ip, "Failed to forward response to client");
            break;
        }
        total_bytes += bytes_sent;
    }

    logger.log_access(client_ip, "RESPONSE", "", 200, total_bytes);
}
int main() {
    try {
        Proxy proxy(8080);
        proxy.start();
    } catch (const std::exception& e) {
        std::cerr << "Fatal error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

